### PR TITLE
fix: export as named exports `default` and `rateLimit`

### LINKED ISSUE #280: TypeError: (0 , express_rate_limit_1.default) is not a function
### Description

my version is 6.0.2, but when i use in nest, it occur an error:

(node:5062) UnhandledPromiseRejectionWarning: TypeError: (0 , express_rate_limit_1.default) is not a function

### Library version

6.0.2

### Node version

14.17.6

### Typescript version (if you are using it)

_No response_

### Module system

ESM

### PR DESCRIPTION
**Pull request by gamemaker1**
## Related Issues

- Fixes #280

### Added

- Export the middleware as `module.exports.default` in CommonJS code.
- Export the middleware as the named export `rateLimit` in both CommonJS and ESM.
- External tests for the named import.

## Checklist

- [x] The issues that this PR fixes/closes have been mentioned above.
- [x] What this PR adds/changes/removes has been explained.
- [x] All tests (`npm test`) pass.
- [x] All added/modified code has been commented, and
      methods/classes/constants/types have been annotated with TSDoc comments.
- [x] If a new feature has been added or a bug has been fixed, tests have been
      added for the same.

### CODE REVIEW

**Review by nfriedly — APPROVED** [review]
💯 Awesome 💯 

I'd be interested in hearing what @DavideViolante says, but it looks good to me.

Also, I feel a little bad for how many variations of that test you've had to create, but I do think that we're getting a very reliable and well-tested library out of the deal. ❤️

File: source/index.ts [hunk 1]
```diff
@@ -3,7 +3,15 @@
...
+
+// Export the rateLimit function as a default export
 // https://github.com/timocov/dts-bundle-generator/issues/182
 // eslint-disable-next-line unicorn/prefer-export-from
 export default rateLimit
+
+// Also export it as a named export if the default export does not work
+// (see https://github.com/nfriedly/express-rate-limit/issues/280)
+
+export { default as rateLimit } from './lib.js'
```
[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts:17]: Ahh the linter must have autofixed that. The recommended style of re export is to do an export from (as it is right now), and I prefer that style too.

I had actually kept it `export { rateLimit }` as there was an import already, but the linter always wins 😅

[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts:17]: Once the issue in the dts-bundle-generator library is fixed, I'll make that `export { default, default as rateLimit } from './lib.js'`

[the remark this answers is not in the document]
Comment (nfriedly) [reply source/index.ts:17]: Can you link to the issue here?

[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts:17]: It's just above it

[the remark this answers is not in the document]
Comment (DavideViolante) [reply source/index.ts:17]: Which linting rule doesn't allow you to export "normally"? It seems weird to me too.

[the remark this answers is not in the document]
Comment (nfriedly) [reply source/index.ts:17]: Sorry, I meant the issue in the dts-bundle-generator library.

[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts:17]: It's the one just below, disabled for the default export (unicorn/prefer-export-from)

[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts:17]: > Sorry, I meant the issue in the dts-bundle-generator library.

Yes, that's on line 10, issue 182 on that repo

[the remark this answers is not in the document]
Comment (nfriedly) [reply source/index.ts:17]: Oh, I see, https://github.com/timocov/dts-bundle-generator/issues/182

[the remark this answers is not in the document]
Comment (nfriedly) [reply source/index.ts:17]: And, yeah, this kind of makes sense for the lint rule: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-export-from.md

**Comment by DavideViolante** [comment]
I think it's pretty good. I've only one doubt in case someone import rateLimit in different ways from different files in same project (not on purpose). _I think_ it would create multiple instances of the lib and it would cause some unexpected behaviour.

In general I would merge but there are some things that still sounds weird to me, such as the footer "hack" for esbuild (there must be a way to avoid that and create a cjs esm lib).
In future I would also improve the local tests that still use a "old" bash script. The remote (tests on github actions) and local tests should be the same. Also it seems weird to me testing the different ways to import a library (I imagine that once the lib is bundled correctly, no tests are needed for different ways of import since the bundle is created by esbuild that is tested itself).

File: source/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts:17]: Ok that issue should be fixed by https://github.com/timocov/dts-bundle-generator/issues/183, which landed in 6.4.0 - I'll change the export now

**Comment by gamemaker1** [comment]
> I think it would create multiple instances of the lib and it would cause some unexpected behaviour.

Importing a function should not create multiple 'instances of the library' or cause undefined behaviour. I believe you might be referring to the [dual package hazard](https://nodejs.org/api/packages.html#dual-package-hazard), which exists in most dual-export libraries and occurs if the package is imported as a CommonJS in one place and ESM in other. We follow the [second approach](https://nodejs.org/api/packages.html#approach-2-isolate-state) specified in the Node documentation to minimize the risk of this occurring.

> The footer "hack" for esbuild

The footer part is only required so that CommonJS users can default-import the library instead of doing something like this:

```ts
const { default: rateLimit } = require('express-rate-limit')
```

ESBuild actually outputs 100% correct javascript, but that means the `export default ...` gets converted to `module.exports.default =  ...` instead of `module.exports = ...`. The footer makes sure both exports are made (`module.exports = rateLimit; module.exports.default = rateLimit`).

> I would improve the local tests that still use a "old" bash script (I think?)

Could you please elaborate?

> The remote (tests on github actions) and local tests should be the same

They are the same:

| GitHub Action | Command |
| ------------- | ------- |
| [CI / Lint](https://github.com/nfriedly/express-rate-limit/blob/master/.github/workflows/ci.yaml#L9-L26)     | [`npm run lint`](https://github.com/nfriedly/express-rate-limit/blob/master/package.json#L57-L59) |
| [CI / Test (Library)](https://github.com/nfriedly/express-rate-limit/blob/master/.github/workflows/ci.yaml#L27-44) | [`npm run test:lib`](https://github.com/nfriedly/express-rate-limit/blob/master/package.json#L63) |
| [CI / Test (External)](https://github.com/nfriedly/express-rate-limit/blob/master/.github/workflows/ci.yaml#L45-L76) | [`npm run test:ext`](https://github.com/nfriedly/express-rate-limit/blob/master/package.json#L64) |
| [CI / Publish](https://github.com/nfriedly/express-rate-limit/blob/master/.github/workflows/ci.yaml#L77-L105) | `npm publish` |

The [`npm test`](https://github.com/nfriedly/express-rate-limit/blob/master/package.json#L65) command runs the first 3 - however, it is split up in the CI so we can run them on different platforms too.

> Testing the different ways to import a library (I imagine that once the lib is bundled correctly, no tests are needed for different ways of import since the bundle is created by esbuild that is tested itself)

I think of the import tests as kind of e2e tests. When they break, they warn us that the change will make the library unusable for the thousands of developers using the library.

Also, ESBuild bundles the library very quickly, and very correctly - it is us that add the weird footer to support default imports in CommonJS.

---

I am merging this and releasing it as 6.1.0 (minor bump as it adds a 'feature', the named export of the `rateLimit` function) 🎉 

Thanks for your help on fixing the issue @DavideViolante!

File: source/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gamemaker1) [reply source/index.ts]: Done in c97488c!

**Pull request merged**